consensus/validator/validator.go

```diff
@@ -0,0 +1,198 @@
+package validator
+
+import (
+	"fmt"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/builder"
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/mempool"
+)
+
+const defaultTxnPoolSize int = 1024
+
+// Validator is used to validate new proposals. There are two potential flows, and functions must be called in order:
+// Flow 1) Non-empty proposal - ProposalInit(), BlockInfo(), TransactionBatch(), ProposalCommitment(), ProposalFin()
+// Flow 2) Empty proposal - ProposalInit(), ProposalCommitment(), ProposalFin()
+//
+//go:generate mockgen -destination=../mocks/mock_validator.go -package=mocks github.com/NethermindEth/juno/consensus/validator Validator
+type Validator[V types.Hashable[H], H types.Hash, A types.Addr] interface {
+	// ProposalInit initialises the pending state according to the proposal
+	ProposalInit(pInit *types.ProposalInit) error
+
+	// BlockInfo sets the pending header according to the proposal header
+	BlockInfo(blockInfo *types.BlockInfo)
+
+	// TransactionBatch executes the provided transactions, and stores the result in the pending state
+	TransactionBatch(txn []types.Transaction) error
+
+	// ProposalCommitment checks the set of proposed commitments against those generated locally.
+	ProposalCommitment(proposalCommitment *types.ProposalCommitment) error
+
+	// ProposalFin compares the provided commitment with that generated locally
+	ProposalFin(proposalFin types.ProposalFin) error
+}
+
+type validator[V types.Hashable[H], H types.Hash, A types.Addr] struct {
+	builder           *builder.Builder    // Builder manages the pending block and state
+	txnPool           []types.Transaction // Txns may not arrive in order
```

Comment:
For validator, transactions should arrive in order, because we already have the sequence number for each proposal part. Part re-ordering to match the sequence number should be handled by the P2P layer.